lib/type_check/internals/parser.ex

```diff
@@ -0,0 +1,268 @@
+  """
+  @spec fetch_type(module() | binary(), atom(), arity()) ::
+          {:error, String.t()} | {:ok, any(), list()}
+  def fetch_type(module, type, arity) do
+    case fetch_types(module, type) do
+      {:ok, []} -> {:error, "cannot find type with the given name"}
+      {:ok, types} -> fetch_type(types, arity)
+      {:error, err} -> {:error, err}
+    end
+  end
+
+  @spec fetch_type(list(), arity()) :: {:error, String.t()} | {:ok, any(), list()}
+  defp fetch_type(types, arity) when is_list(types) do
+    case types |> Enum.find(fn {_, {_, _, vars}} -> length(vars) == arity end) do
+      {_, {_, spec, vars}} -> {:ok, spec, vars}
+      nil -> {:error, "cannot find type with the given arity"}
+    end
+  end
+
+  @doc """
+  Get list of all types from the module matching the given name.
+
+  If no types found, empty list is returned.
+  It is possible for list to contain multiple types if there are multiple type
+  definitions with the same name but different amount of generic arguments.
+  """
+  @spec fetch_types(module() | binary() | list(), atom()) :: {:error, String.t()} | {:ok, list()}
+  def fetch_types(module, type) when is_atom(module) or is_binary(module) do
+    case Code.Typespec.fetch_types(module) do
+      {:ok, types} -> fetch_types(types, type)
+      :error -> {:error, "cannot fetch types from the module"}
+    end
+  end
+
+  def fetch_types(types, type) when is_list(types) do
+    filtered = types |> Enum.filter(fn {_, {name, _, _}} -> name == type end)
+    {:ok, filtered}
+  end
+
+  @doc """
+  Extract module name, function name, and arguments from quoted expression of a function call.
+
+  ## Examples
+
+      iex> import TypeCheck.Internals.Parser
+      iex> ast_to_mfa(quote do: abs(1))
+      {Kernel, :abs, [1]}
+      iex> ast_to_mfa(quote do: Kernel.abs(1))
+      {Kernel, :abs, [1]}
+      iex> ast_to_mfa(quote do: Date.diff(1, 2))
+      {Date, :diff, [1, 2]}
+      iex> alias Date, as: D
+      iex> ast_to_mfa(quote do: D.diff(1, 2))
+      {Date, :diff, [1, 2]}
+  """
+  @spec ast_to_mfa(Macro.t()) :: {module(), atom(), [Macro.t()]}
+  def ast_to_mfa({func, [context: _, import: module], args}) do
+    {module, func, args}
+  end
+
+  def ast_to_mfa({target, _, args}) do
+    case target do
+      {:., _, [{:__aliases__, [alias: false], [mod]}, func]} -> {elixir_module(mod), func, args}
+      {:., _, [{:__aliases__, [alias: mod], _}, func]} -> {mod, func, args}
+      {:., _, [{:__aliases__, _, [mod]}, func]} -> {elixir_module(mod), func, args}
+      {:., _, [mod, func]} when is_atom(mod) -> {mod, func, args}
+      _ -> raise("cannot infer function")
+    end
+  end
+
+  def ast_to_mfa(_), do: raise("not a function call")
+
+  defp elixir_module(module), do: String.to_atom("Elixir.#{module}")
+
+  @doc """
+  Convert the raw spec extracted by `fetch_spec/3` into type_check type.
+
+  The function is optimistic. If the type is not known, it assumes `any()`.
```

Comment:
I'm not entirely convinced that we should always be optimistic rather than being explicit about what we do and don't support.

Maybe we should add an option for the user to decide between them.
Something for a later version; please remind me that we add an issue for this.

Reply:
That's why I've added `ctx.default`. I also thought about adding a `ctx.raise` flag. Neither is exposed to the user just yet, though.